src/ctap/storage.rs

```diff
@@ -214,310 +127,281 @@ impl PersistentStore {
-            self.set_aaguid(key_material::AAGUID).unwrap();
+        self.load_attestation_data_from_firmware()?;
+        if self.store.find_handle(key::AAGUID)?.is_none() {
+            self.set_aaguid(key_material::AAGUID)?;
         }
+        Ok(())
     }
 
     // TODO(jmichel): remove this function when vendor command is in place.
     #[cfg(not(test))]
-    fn load_attestation_data_from_firmware(&mut self) {
+    fn load_attestation_data_from_firmware(&mut self) -> Result<(), Ctap2StatusCode> {
         // The following 2 entries are meant to be written by vendor-specific commands.
-        if self.store.find_one(&Key::AttestationPrivateKey).is_none() {
-            self.set_attestation_private_key(key_material::ATTESTATION_PRIVATE_KEY)
-                .unwrap();
+        if self
+            .store
+            .find_handle(key::ATTESTATION_PRIVATE_KEY)?
+            .is_none()
+        {
+            self.set_attestation_private_key(key_material::ATTESTATION_PRIVATE_KEY)?;
         }
-        if self.store.find_one(&Key::AttestationCertificate).is_none() {
-            self.set_attestation_certificate(key_material::ATTESTATION_CERTIFICATE)
-                .unwrap();
+        if self
+            .store
+            .find_handle(key::ATTESTATION_CERTIFICATE)?
+            .is_none()
+        {
+            self.set_attestation_certificate(key_material::ATTESTATION_CERTIFICATE)?;
         }
+        Ok(())
     }
 
+    /// Returns the first matching credential.
+    ///
+    /// Returns `None` if no credentials are matched or if `check_cred_protect` is set and the first
+    /// matched credential requires user verification.
     pub fn find_credential(
         &self,
         rp_id: &str,
         credential_id: &[u8],
         check_cred_protect: bool,
     ) -> Result<Option<PublicKeyCredentialSource>, Ctap2StatusCode> {
-        let key = Key::Credential {
-            rp_id: Some(rp_id.into()),
-            credential_id: Some(credential_id.into()),
-            user_handle: None,
-        };
-        let entry = match self.store.find_one(&key) {
-            None => return Ok(None),
-            Some((_, entry)) => entry,
-        };
-        debug_assert_eq!(entry.tag, TAG_CREDENTIAL);
-        let result = deserialize_credential(entry.data);
-        debug_assert!(result.is_some());
-        let user_verification_required = result.as_ref().map_or(false, |cred| {
-            cred.cred_protect_policy == Some(CredentialProtectionPolicy::UserVerificationRequired)
+        let mut iter_result = Ok(());
+        let iter = self.iter_credentials(&mut iter_result)?;
+        // TODO(reviewer): Should we return an error if we find more than one matching credential?
+        // We did not use to in the previous version (panic in debug mode, nothing in release mode)
+        // but I don't remember why. Let's document it.
+        let result = iter.map(|(_, credential)| credential).find(|credential| {
```

Comment:
Rust / style question: Can we replace
`map(|(_, credential)| credential).find(|credential|`
with
`find(|(_, credential)|` ?

Reply:
There shouldn't be any difference. On a simple example we see that they get compiled to the same: https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=f3783785dde9efa7e982468d4792bfab (when showing assembly).

Reply:
Nice, thanks for the effort.